### PR TITLE
Fix updating v3 npm lockfiles

### PR DESCRIPTION
**Pull request by deivid-rodriguez**
Fixes https://github.com/dependabot/dependabot-core/issues/6270.

### CODE REVIEW

**Review by jurre — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
I [tried this on a real repo](https://github.com/dsp-testing/test-dependabot-npm-lockfile-v3-workspace/pull/1) and it worked fine, except for hitting this npm issue: https://github.com/npm/cli/issues/2700. I think the new PRs are still in a much better place that without updating the lockfile at all, so good to go I think!

**Comment by deivid-rodriguez** [comment]
https://github.com/npm/cli/issues/5470 also seems related. In any case, the better lockfile seems the one that includes "name" fields, and that's what Dependabot is generating. Whatever it is, it seems like a package manager issue.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb [hunk 1]
```diff
@@ -155,31 +155,37 @@ def shrinkwrap_dependencies
...
+          end
 
-              dependency_args = {
-                name: name,
-                version: semver_version_for(details["version"]),
-                package_manager: "npm_and_yarn",
-                requirements: []
-              }
+          dependencies.each do |name, details|
+            next if name.empty? # v3 lockfiles include an empty key holding info of the current package
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:164]: So if I'm reading correctly the code that changed is really only the lines above... below is just whitespace changes due to nesting within the `each do` block...??

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb [hunk 2]
```diff
@@ -155,31 +155,37 @@ def shrinkwrap_dependencies
         def recursively_fetch_npm_lock_dependencies(object_with_dependencies)
           dependency_set = Dependabot::NpmAndYarn::FileParser::DependencySet.new
 
-          object_with_dependencies.
-            fetch("dependencies", {}).each do |name, details|
-              next unless semver_version_for(details["version"])
+          dependencies = object_with_dependencies["dependencies"]
+          dependencies ||= object_with_dependencies.fetch("packages", {}).transform_keys do |name|
+            name.delete_prefix("node_modules/")
```
Comment (jeffwidman) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:160]: Why does this get deleted? Also why is is fetching `"packages"` when before it didn't?

**Review by jeffwidman — COMMENTED** [review]
Can you provide a little more context for those unfamiliar with the changes that landed in the new `npm` lockfile format?

I read the linked issue, but afraid I still don't quite follow...

> I think the new PRs are still in a much better place that without updating the lockfile at all, so good to go I think!

Yeah, as long as the logic is correct for what we do, doesn't seem like we should block this on upstream bugs unless they impact our behavior...

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb (hunk 1 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:164]: Yes, this is one PR to review with `w=1` :)

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb (hunk 2 above)
Comment (deivid-rodriguez) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:160]: This is the v3 format, which no longer provides `dependencies` key, and prepends `node_modules/` to packages names.

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb (hunk 1 above)
Comment (jeffwidman) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser/lockfile_parser.rb:164]: oh right, you can do that in the github UI, awesome! makes it much, much simpler

**Review by jeffwidman — APPROVED** [review]

**Comment by deivid-rodriguez** [comment]
Merging since missing lockfile updates is worse than some potential flip-flop on lockfile format due to npm issues.

**Pull request merged**